setup.go

```diff
@@ -27,11 +27,19 @@ type config struct {
 	ifName        string
 	localIPs      []net.IP
 	setupIPTables bool
+	skipTeardown  bool
 }
 
 func setup(c *caddy.Controller) error {
-	cfg, err := parseConfig(dnsserver.GetConfig(c))
-	if err != nil {
+
+	cfg := getDefaultCfg()
+	if cfg.parsePlgConfig(c) != nil {
+		log.Errorf("Error while parsing plugin config")
+		return plugin.Error("nodecache", c.ArgErr())
+	}
+
+	if cfg.parseSrvConfig(dnsserver.GetConfig(c)) != nil {
+		log.Errorf("Error while parsing server config")
```

Comment:
splitting it here to parse the server config independent of the plugin config (the single statement line)